lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -1343,14 +1372,56 @@ struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
 // This is the main entrypoint for the lowering pass.
 void LowerTypesPass::runOnOperation() {
   std::vector<Operation *> ops;
-  llvm::for_each(getOperation().getBody()->getOperations(),
-                 [&](Operation &op) { ops.push_back(&op); });
+  // Record of each module, to the map of NonLocalAnchors which need to be
+  // updated with the lowered inner_sym name.
+  DenseMap<Operation *, DenseMap<StringAttr, StringAttr>>
+      modToNlaNameToNewSymMap;
+  // Map of name of the NonLocalAnchor to the operation.
+  DenseMap<StringAttr, Operation *> nlaMap;
+  // Record all operations in the circuit.
+  llvm::for_each(getOperation().getBody()->getOperations(), [&](Operation &op) {
+    ops.push_back(&op);
+    // Creating a map of all ops in the circt, but only modules are relevant.
+    modToNlaNameToNewSymMap.insert(
+        std::make_pair<Operation *, DenseMap<StringAttr, StringAttr>>(
+            &op, /*An empty map*/ DenseMap<StringAttr, StringAttr>()));
+    // Record the NonLocalAnchor and its name.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op))
+      nlaMap[nla.sym_nameAttr()] = nla;
+  });
 
+  // Each parallel module worker updates the map to the corresponding module.
+  // After this statement runs, modToNlaNameToNewSymMap[ops[index]] will be
+  // updated with all NonLocalAnchors that need to be updated with the lowerd
+  // symbols.
   mlir::parallelForEachN(&getContext(), 0, ops.size(), [&](auto index) {
```

Comment:
In fact, you have an index, so you can preallocate a vector of per-module data and index it safely.